integration_test.go

```diff
@@ -0,0 +1,231 @@
+	duration, err := time.ParseDuration(r.FormValue("duration"))
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+	time.Sleep(duration)
+	fmt.Fprintf(
+		w, "Slept for %2v seconds. Response from pid %d.\n",
+		duration.Seconds(),
+		os.Getpid(),
+	)
+}
+
+// createIntegrationServer will re-run the test binary that is currently running,
+// but only execute the "TestIntegrationServer" test, which will stand up an
+// 'endless' server which can be used to test signal handling
+func createIntegrationServerTest() *exec.Cmd {
+	cmd := exec.Command(os.Args[0], "-test.run=TestIntegrationServer")
+
+	// Set the feature toggle for the integration test server
+	cmd.Env = []string{IntegrationServerFeatureToggle + "=1"}
+
+	return cmd
+}
+
+func TestIntegrationRestart(t *testing.T) {
+
+	intServer := createIntegrationServerTest()
+	serverMessages := make(chan string) // channel to receive notifications from the integration server
+	// Start test integration server
+	require.NoError(t, startIntegrationServer(intServer, serverMessages), "starting integration test server")
+	parentPid := intServer.Process.Pid
+	defer teardownServer(parentPid)
+	go intServer.Wait() // when this ends the parent process should be over and pipe clean up will happen
+
+	// The pid is printed out from the integation test server once it has fully started up
+	// Check for this message so we can start sending requests/signals to the server
+	msg, msgChannelOpen := <-serverMessages
+	require.True(t, msgChannelOpen, "Unexpected close of server messages channel")
+	msgPid, _ := strconv.Atoi(msg)
+	require.Equal(t, parentPid, msgPid, "Unexpected message down server messages channel")
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the server a chance to bind to port
+
+	// Send a long running request to the test server
+	longReqWait := &sync.WaitGroup{}
+	longReqWait.Add(1)
+	var (
+		longResp     *http.Response
+		longReqError error
+	)
+	go func() {
+		longResp, longReqError = sendIntegrationTestRequest(2 * time.Second)
+		longReqWait.Done()
+	}()
+
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the request a chance to get handled
+
+	// Send a signal to the test server to restart
+	require.NoError(t, intServer.Process.Signal(syscall.SIGHUP), "send signal to restart integration test server")
+
+	// The child server should also print it's pid out once it's started up, so look for this
+	msg, msgChannelOpen = <-serverMessages
+	require.True(t, msgChannelOpen, "Unexpected close of server messages channel")
+	msgPid, _ = strconv.Atoi(msg)
+	require.NotZero(t, msgPid, "Unexpected message down server messages channel")
+	childPid := msgPid
+	defer teardownServer(childPid) // set up the child server to be torn down on test exit
+
+	quickResp, quickReqError := sendIntegrationTestRequest(0 * time.Second)
+	require.NoError(t, quickReqError, "sending quick request to child integration test server")
+	require.Equal(t,
```

Comment:
excessive wrapping, which IMO makes this harder to read.

Reply:
There were quite a few long messages explaining why the error occurred with the requires, so I took the decision to wrap all of them for consistency, which do you prefer:
* consistency so all the requires look the same
* or only wrap the long lines
* or don't wrap any

Reply:
I guess there a few points
1. The messages should be short and snappy.
2. Long vs short var names are making your problem worse. e.g. ```parentPid``` vs ```pid```, ```quickResp``` vs ```qr```
2. I would prefer non-wrapping personally, we'll open it up to the wider group.
As some examples:
```go
require.Equal(t, http.StatusOK, qr.StatusCode, "long running request")
require.NotContains(t, string(rwBody), strconv.Atoi(pid), "quick response contained pid")
```

Reply:
Fair point, will re-write